// git-neo/lib/ai-commit.js
const { GoogleGenerativeAI } = require('@google/generative-ai');
const chalk = require('chalk');
const ora = require('ora');

class AICommitGenerator {
    constructor(config = null) {
        this.config = config;

        // Priority: 1. Config, 2. Environment variable
        this.apiKey = (config && config.aiApiKey) ||
            process.env.GEMINI_API_KEY ||
            process.env.GOOGLE_API_KEY;

        this.model = (config && config.aiModel) || 'gemini-pro';
        this.genAI = null;

        if (this.apiKey) {
            this.genAI = new GoogleGenerativeAI(this.apiKey);
        }
    }

    isAvailable() {
        // Check if AI is enabled in config (if config exists)
        if (this.config && this.config.aiEnabled === false) {
            return false;
        }
        return !!this.genAI;
    }

    async generateCommitMessage(diff) {
        if (!this.isAvailable()) {
            return {
                success: false,
                message: 'AI not configured. Run "git-neo setup" to configure AI.'
            };
        }

        const spinner = ora('Generating AI commit message...').start();

        try {
            const model = this.genAI.getGenerativeModel({ model: this.model });

            const prompt = `You are an expert Git commit message writer. Analyze the following git diff and generate a concise, meaningful commit message.

Rules:
1. Use conventional commit format (feat:, fix:, docs:, style:, refactor:, test:, chore:)
2. Keep it under 72 characters
3. Be specific about what changed
4. Use imperative mood (e.g., "add" not "added")
5. Don't include file names unless critical
6. Focus on the WHY and WHAT, not the HOW

Git Diff:
${diff}

Generate ONLY the commit message, nothing else. No explanations, no markdown, just the commit message.`;

            const result = await model.generateContent(prompt);
            const response = await result.response;
            const text = response.text().trim();

            spinner.succeed(chalk.green('✓ AI commit message generated'));

            return {
                success: true,
                message: text
            };

        } catch (error) {
            spinner.fail(chalk.red('✗ AI generation failed'));

            if (error.message && error.message.includes('API_KEY')) {
                return {
                    success: false,
                    message: 'Invalid API key. Run "git-neo setup" to reconfigure.'
                };
            }

            return {
                success: false,
                message: error.message || 'AI generation failed'
            };
        }
    }

    async generateMultipleOptions(diff, count = 3) {
        if (!this.isAvailable()) {
            return {
                success: false,
                message: 'AI not configured. Run "git-neo setup" to configure AI.'
            };
        }

        const spinner = ora(`Generating ${count} AI commit message options...`).start();

        try {
            const model = this.genAI.getGenerativeModel({ model: this.model });

            const prompt = `You are an expert Git commit message writer. Analyze the following git diff and generate ${count} different commit message options.

Rules:
1. Use conventional commit format (feat:, fix:, docs:, style:, refactor:, test:, chore:)
2. Keep each under 72 characters
3. Be specific about what changed
4. Use imperative mood (e.g., "add" not "added")
5. Provide different perspectives/focuses for each option
6. Focus on the WHY and WHAT, not the HOW

Git Diff:
${diff}

Generate ONLY ${count} commit messages, one per line. No numbering, no explanations, no markdown.`;

            const result = await model.generateContent(prompt);
            const response = await result.response;
            const text = response.text().trim();

            // Split into lines and clean up
            const messages = text
                .split('\n')
                .map(line => line.trim())
                .filter(line => line.length > 0 && !line.match(/^\d+\./)) // Remove numbered lines
                .slice(0, count); // Ensure we only get the requested count

            spinner.succeed(chalk.green(`✓ Generated ${messages.length} AI commit message options`));

            return {
                success: true,
                messages: messages
            };

        } catch (error) {
            spinner.fail(chalk.red('✗ AI generation failed'));

            return {
                success: false,
                message: error.message || 'AI generation failed'
            };
        }
    }

    getSetupInstructions() {
        return `
${chalk.bold.cyan('🤖 AI Commit Message Setup')}

To use AI-generated commit messages:

${chalk.bold('Option 1: Configure during setup (Recommended)')}
   ${chalk.gray('git-neo setup')}
   Then choose to enable AI commits and enter your API key

${chalk.bold('Option 2: Set Environment Variable')}
   
   ${chalk.bold('Windows (PowerShell):')}
   ${chalk.gray('$env:GEMINI_API_KEY="your-api-key-here"')}
   
   ${chalk.bold('Windows (CMD):')}
   ${chalk.gray('set GEMINI_API_KEY=your-api-key-here')}
   
   ${chalk.bold('Linux/Mac:')}
   ${chalk.gray('export GEMINI_API_KEY="your-api-key-here"')}

${chalk.bold('Get API Key:')}
   Visit: ${chalk.cyan('https://makersuite.google.com/app/apikey')}
   Click "Create API Key" and copy it

${chalk.yellow('Note:')} The API key is free for personal use with rate limits.
`;
    }
}

module.exports = AICommitGenerator;
